pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/RealtimeLuceneTextIndex.java

```diff
@@ -109,27 +112,46 @@ public ImmutableRoaringBitmap getDictIds(String searchQuery) {
   @Override
   public MutableRoaringBitmap getDocIds(String searchQuery) {
     MutableRoaringBitmap docIDs = new MutableRoaringBitmap();
-    Collector docIDCollector = new RealtimeLuceneDocIdCollector(docIDs);
-    IndexSearcher indexSearcher = null;
-    try {
-      Query query = new QueryParser(_column, _analyzer).parse(searchQuery);
-      indexSearcher = _searcherManager.acquire();
-      indexSearcher.search(query, docIDCollector);
-      return getPinotDocIds(indexSearcher, docIDs);
-    } catch (Exception e) {
-      LOGGER
-          .error("Failed while searching the realtime text index for column {}, search query {}, exception {}", _column,
-              searchQuery, e.getMessage());
-      throw new RuntimeException(e);
-    } finally {
+    RealtimeLuceneDocIdCollector docIDCollector = new RealtimeLuceneDocIdCollector(docIDs);
+    // A thread interrupt during indexSearcher.search() can break the underlying FSDirectory used by the IndexWriter
+    // which the SearcherManager is created with. To ensure the index is never corrupted the search is executed
+    // in a child thread and the interrupt is handled in the current thread by canceling the search gracefully
+    Callable<MutableRoaringBitmap> searchCallable = () -> {
+      IndexSearcher indexSearcher = null;
       try {
-        if (indexSearcher != null) {
-          _searcherManager.release(indexSearcher);
+        Query query = new QueryParser(_column, _analyzer).parse(searchQuery);
+        indexSearcher = _searcherManager.acquire();
+        indexSearcher.search(query, docIDCollector);
+        return getPinotDocIds(indexSearcher, docIDs);
+      } finally {
+        try {
+          if (indexSearcher != null) {
+            _searcherManager.release(indexSearcher);
+          }
+        } catch (Exception e) {
+          LOGGER.error(
+              "Failed while releasing the searcher manager for realtime text index for column {}, exception {}",
+              _column, e.getMessage());
+        }
+      }
+    };
+    FutureTask<MutableRoaringBitmap> searchTask = new FutureTask<>(searchCallable);
+    Thread searcherThread = new Thread(searchTask);
+    searcherThread.start();
+    try {
+      while (!searchTask.isDone()) {
```

Comment:
Could the same behavior be achieved by using `Object.wait()`?

Essentially you create a dummy object, say "LuceneNotifier" and pass it to the search callable. The search callable will call `luceneNotifier.notify()` in the finally block. And the while loop would be replaced with something like this:

```
try {
  luceneNotifier.wait();
  // code reaches here if notify was call after success/failure
} catch (InterruptedException e) {
  // this gets called if the thread was interrupted
  docIdCollector.markShouldCancel();
}
```

Reply:
ahh, I don't think I can use `.notify()` in the finally block since it's likely to get called before the `.wait()` is even reached - and moving the `.wait()` into a loop as required ends up in about the same place we started

But I can simplify it with just `future.get()` and catch the interrupt, so thank you 😄

Reply:
My bad I miscommunicated in my comment: what I really meant was that we could use a signaling mechanism instead of the while loop. A `CountDownLatch` would be better than using Object wait and notify.

But yeah simply using future.get is even better :D